### PR TITLE
define Constructor for Source, Sink and Store

### PR DESCRIPTION
**Pull request by elf-pavlik**
it tries to address #6 and #5 and looks at current implementations in @rdfjs orga

### CODE REVIEW

**Comment by elf-pavlik** [comment]
@jacoscaz this might affect https://github.com/beautifulinteractions/node-quadstore#rdfstore-class which uses `dataFactory` in `options` and has it as second argument.

Would you see possibility of passing `abstractLevelDOWN` in the `options` object?
```js
RdfStore({
  dataFactory,
  abstractLevelDOWN: memdown()
})
```

Or maybe using partial application?
```js
RdfStore(memdown())({ dataFactory })
```

**Comment by elf-pavlik** [comment]
@rubensworks it also seems to affect https://github.com/rdfjs/rdfxml-streaming-parser.js#configuration

I see it also uses `dataFactory` which means currently even different repositories in @rdfjs orga don't have compatible constructors.

**Comment by jacoscaz** [comment]
@elf-pavlik no issues for us! We'll just release a new major version including this and other breaking changes (if any).

File: index.html [hunk 1]
```diff
@@ -140,6 +140,12 @@ <h4>Optional Events</h4>
       <h3><dfn>Source</dfn> interface</h3>
 
       <pre class="idl">
+      interface SourceOptions {
+        attribute DataFactory? factory;
```
Comment (rubensworks) [inline index.html]: I wasn't around for the discussion in #6, but I actually have a slight preference for using `dataFactory` as variable name.

The reason for this is that implementations may accept other kinds of internal _factories_ as well, for example a `nodeFactory`, `numberFactory`, `errorFactory`, ...
For this reason, the `factory` name may be too generic.

Eitherway, it should not be a problem to support for `factory` and `dataFactory` for ensuring backwards-compability.

Comment (elf-pavlik) [reply index.html]: @bergos @RubenVerborgh I think you both use `factory` in N3.js and parser-jsonld, personally I might also have slight preference for `dataFactory` which makes things crystal clear including code like
```js
import dataFactory from '@rdfjs/data-model'

// ....

const parser = new Parser({ dataFactory })
```

As mentioned above `node-quadstore` also uses `dataFactory` https://github.com/rdfjs/stream-spec/pull/10#issuecomment-469855223

Comment (RubenVerborgh) [reply index.html]: Preference for `dataFactory` confirmed; will support both in N3.js if adjusted.

**Review by rubensworks — APPROVED** [review]

**Comment by elf-pavlik** [comment]
Thanks for review @rubensworks 
@bergos @blake-regalia @RubenVerborgh I know you all also maintain streaming parsers, I hope you will have chance to review this PR

File: index.html [hunk 2]
```diff
@@ -179,6 +200,17 @@ <h3><dfn>Sink</dfn> interface</h3>
...
       </p>
+      <p>
+        <dfn>SinkOptions</dfn> Sink constructor can take <code>options</code> object as parameter.
+      </p>
+      <div data-dfn-for="SinkOptions">
+        <p>
+          <dfn>baseIRI</dfn> Base IRI used to resolve or create relative IRIs.
+        </p>
+        <p>
+          <dfn>dataFactory</dfn> DataFactory that Sink will use if <code>import()</code> returns <code>Stream&lt;Quad&gt;</code>.
```
Comment (RubenVerborgh) [inline index.html]: Not sure I get this… why would Sink need that?

**Review by RubenVerborgh — APPROVED** [review]

File: index.html (hunk 2 above)
Comment (rubensworks) [reply index.html]: This is needed if the stream is a text stream, for example if the sink is a parser.

Comment (RubenVerborgh) [reply index.html]: But then the explanation might be wrong; it's not about the output, but about the input then?

(Also confusing that we use Source and Sink also for text; we probably want something else.)

Comment (rubensworks) [reply index.html]: Agreed, see #2.

**Comment by bergos** [comment]
I used `factory` in the packages, cause I would use a single factory object, which provides all factory methods from the data-model, dataset and upcoming specs (prefix could be such a candidate), in my code. With prefix and CURIE support, the factory would go more in the direction of an environment, like defined in the [RDF Interfaces spec](https://www.w3.org/TR/rdf-interfaces/#rdf-environment). With that in mind, I'm a little bit in favor for `factory`. Technical it's not a problem to pass the factory to multiple factory options, so if the majority is for `dataFactory`, I would also add support for both.

File: index.html (hunk 2 above)
Comment (elf-pavlik) [reply index.html]: Few lines below we have

> Typical use cases
> Parser:
> Stream&lt;Quad&gt; import(Stream stream)

`import()` takes stream of text as argument and returns stream of quads. How would you explain that `dataFactory` will get used to construct those quads in the stream returned by `import()` ?

Comment (elf-pavlik) [reply index.html]: `output` variable in this example: https://github.com/rdfjs/parser-jsonld#example

**Comment by elf-pavlik** [comment]
> I used `factory` in the packages, cause I would use a single factory object, which provides all factory methods from the data-model, dataset and upcoming specs (prefix could be such a candidate), in my code.

I think it may make sense to have `factory` print out deprecation warring, otherwise developers might get used to non standard `factory` and get caught by surprise where other constructors don't support it. Even worst if they pass specific `factory` and implementation of S* will use a different default one instead, since standard `dataFactory` was `undefined`.

When we add support for `prefixMap` I think it should come as additional option parameter, just as `baseIRI` or whatever we may need for JSON-LD context for serializers which support compacting. I see difference here between DataFactory and RDFEnvironment which seems to include prefixes.

File: index.html (hunk 2 above)
Comment (RubenVerborgh) [reply index.html]: Okay, works for me.

Comment (RubenVerborgh) [reply index.html]: Just the whole Stream/Stream<> is something we need to fix, but other issue.

**Comment by bergos** [comment]
Maybe it would be easier to define a single `ConstructorOptions` interface, which contains all possible options. For example:

- `Source` could be used directly by a parser which accepts a baseIRI
- `Store` could be just a wrapper around a dataset, parser and serializer, where booth, parser and serializer, accept a baseIRI

I'm not sure if we can think of use cases for all possible combinations as of the time of writing the spec.

**Comment by elf-pavlik** [comment]
I'll make that change tomorrow if no one argues against it.

**Comment by elf-pavlik** [comment]
I have unified all options into single `ConstructorOptions` following @bergos suggestion.

**Comment by elf-pavlik** [comment]
I will give this PR 2-3 more days and merge it. My last change should not affect the 2 approvals it already received.

**Review by bergos — APPROVED** [review]

**Pull request merged**